table.for_each for iteration

ThePhD opened the pull request:
A temporary but powerful solution to the lack of iterators, `table.for_each` lets a person iterate over the keys and values for a specified `sol::table` with each so long as they give it a function which takes a `sol::object key, sol::object value`.

ThePhD committed:
new table.for_each while iterators get sorted out

ThePhD committed:
Fix tabs.

Rapptz commented:
I appreciate the addition but I'm not sure if it's actually worth adding.

I'll add it if you add tests.

ThePhD committed:
Standard-compliant function signature deduction (cannot use set_function<Args...> format as that creates an ambiguity)
state now has for_each (runs on global table like all other functions)
added for_each tests, per @Rapptz request

ThePhD commented:
@Rapptz  All set!

Rapptz commented on sol/table.hpp:
```diff
@@ -98,9 +98,24 @@ class table : public reference {
         push();
-        return lua_rawlen(state(), -1);
+        size_t result = lua_rawlen(state(), -1);
+        pop();
```
What's with the `pop`?

ThePhD replied:
It actually left a `table` instance on the stack, since `lua_rawlen` does not pop the table or push anything on the stack (lua_len does push something onto the stack).
See: http://puu.sh/hOU9T/9efa5d70e9.png

Rapptz commented:
I'm thinking of moving this from a member function to a free function.

```
sol::for_each(table, func);
```

But I guess iterators (which we've discussed extensively) would be a better solution which is hard-ish to do.

ThePhD commented:
It's do-able if you used a `std::shared_ptr` to put the state of the iterator, because that allows the "shallow-copyable" iterator principle the `std::` uses to work. The only semantics that don't work are saving state (e.g. `auto it = table.begin(); auto saved = it; ++it; ++it;/* incrementing either will result in being 2 ahead on both, not 1, skipping a value */` ). I can't think of a way to do "saved position in iteration" because everything is tied to the current lua stack, and there's no way to successfully dupe the stack mid-iteration without _destroying_ the iteration (and causing a segfault if you attempt to keep going after "fixing" the stack).
